File: Algorithm_LevelOne/Longest_Substring_Without_Repeating_Characters/Longest_Substring_Without_Repeating_Characters.py
# 3. Longest Substring Without Repeating Characters Python

class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        lst = []
        longest = 0
        count = 0
        for i in range(len(s)):
            if s[i] in lst:
                if longest < count:
                    longest = count
                index = lst.index(s[i])
                count -= index
                lst = lst[index + 1:]
                lst.append(s[i])
            else:
                lst.append(s[i])
                count += 1
        return count if longest < count else longest
                

File: LeetCode75_LevelTwo/Longest_Substring_Without_Repeating_Characters/Longest_Substring_Without_Repeating_Characters.cc
// 3. Longest Substring Without Repeating Characters C++
#include <string>
#include <vector>

class Solution {
public:
    int lengthOfLongestSubstring(std::string s) {
        std::vector<int> dict(256, -1);
        int maxLen = 0;
        int start = -1;
        for (int i = 0; i < s.size(); i++) {
            if (dict[s[i]] > start)
                start = dict[s[i]];
            dict[s[i]] = i;
            maxLen = std::max(maxLen, i - start);
        }
        return maxLen;
    }
};
